### PR TITLE
v4 - pricelist object + refactor

### PR DESCRIPTION
**Pull request by arik123**
Old pricelist structure:
```json
[
  {
    "sku": "16102;15;w1;pk160",
    "enabled": true,
    "autoprice": false,
    "max": 10,
    "min": 0,
    "intent": 2,
    "buy": {
      "keys": 0,
      "metal": 2.55
    },
    "sell": {
      "keys": 0,
      "metal": 3
    },
    "promoted": 0,
    "group": "all",
    "note": {
      "buy": null,
      "sell": null
    },
    "time": null
  },
  {
    "sku": "16102;15;w1;pk144",
    "enabled": true,
    "autoprice": false,
    "max": 10,
    "min": 0,
    "intent": 2,
    "buy": {
      "keys": 0,
      "metal": 2.55
    },
    "sell": {
      "keys": 0,
      "metal": 3
    },
    "promoted": 0,
    "group": "all",
    "note": {
      "buy": null,
      "sell": null
    },
    "time": null
  },
  {
    "sku": "16102;15;w1;pk122",
    "enabled": true,
    "autoprice": false,
    "max": 10,
    "min": 0,
    "intent": 2,
    "buy": {
      "keys": 0,
      "metal": 3.55
    },
    "sell": {
      "keys": 0,
      "metal": 4.55
    },
    "promoted": 0,
    "group": "all",
    "note": {
      "buy": null,
      "sell": null
    },
    "time": null
  }
]
```

New pricelist structure:
```json
{
    "233;6": {
        "sku": "233;6",
        "name": "Secret Saxton",
        "enabled": true,
        "autoprice": false,
        "max": 1,
        "min": 0,
        "intent": 2,
        "buy": { "keys": 0, "metal": 0.33 },
        "sell": { "keys": 0, "metal": 4 },
        "time": null,
        "promoted": 0,
        "group": "all",
        "note": { "buy": null, "sell": null }
    },
    "30838;6": {
        "sku": "30838;6",
        "name": "Head Prize",
        "enabled": true,
        "autoprice": false,
        "max": 1,
        "min": 0,
        "intent": 2,
        "buy": { "keys": 0, "metal": 0.05 },
        "sell": { "keys": 0, "metal": 1 },
        "time": null,
        "promoted": 0,
        "group": "all",
        "note": { "buy": null, "sell": null }
    },
    "388;5;u57;strange": {
        "sku": "388;5;u57;strange",
        "name": "Strange Terror-Watt Private Eye",
        "enabled": true,
        "autoprice": false,
        "max": 1,
        "min": 0,
        "intent": 2,
        "buy": { "keys": 0, "metal": 1 },
        "sell": { "keys": 100, "metal": 0 },
        "time": null,
        "promoted": 0,
        "group": "all",
        "note": { "buy": null, "sell": null }
    }
}
```

### CODE REVIEW

File: src/schemas/pricelist-json/pricesDataObject.ts [hunk 1]
```diff
@@ -1,16 +1,11 @@
...
-                $ref: '#/definitions/EntryData'
-            }
-        },
-        EntryData: {
+export const pricesDataObject: jsonschema.Schema = {
+    id: 'prices-data-object',
+    type: 'object',
+    patternProperties: {
+        // eslint-disable-next-line prettier/prettier, no-useless-escape
+        '^(\d+);([0-9]|[1][0-5])(;((uncraftable)|(untradable)|(australium)|(festive)|(strange)|((u|pk|td-|c|od-|oq-|p)\d+)|(w[1-5])|(kt-[1-3])|(n((100)|[1-9]\d?))))*?$': {
```
Comment (arik123) [inline src/schemas/pricelist-json/pricesDataObject.ts:8]: I don't thing, that regex is needed here.

Comment (idinium96) [reply src/schemas/pricelist-json/pricesDataObject.ts:8]: so just `^(.)*$` ?

Comment (arik123) [reply src/schemas/pricelist-json/pricesDataObject.ts:8]: I'm not that sure how does the validator work, but if there is just an option to use `string` that would be best

Comment (idinium96) [reply src/schemas/pricelist-json/pricesDataObject.ts:8]: not work with the new structure

**Comment by idinium96** [comment]
Currently if no pricelist were found
image

File: src/classes/Bot.ts [hunk 2]
```diff
@@ -417,7 +417,7 @@ export default class Bot {
                         log.info('Setting up pricelist...');
 
                         void this.pricelist
-                            .setPricelist(!Array.isArray(data.pricelist) ? [] : data.pricelist, this)
+                            .setPricelist(Array.isArray(data.pricelist) ? {} : data.pricelist, this)
```
Comment (idinium96) [inline src/classes/Bot.ts:420]: This is what cause that crash

File: src/classes/Inventory.ts [hunk 3]
```diff
@@ -190,16 +190,14 @@ export default class Inventory {
...
-                    .reduce((total, currentTotal) => (total ? total + currentTotal : currentTotal))
-            );
+            const all = tradableOnly
+                ? Object.keys(this.tradable)
+                : Object.keys(this.tradable).concat(Object.keys(this.nonTradable));
+            return all
+                .filter(e => e.startsWith(sku))
+                .reduce((sum, s) => {
+                    return sum + this.getAmount(s, tradableOnly);
+                }, 0);
```
Comment (idinium96) [inline src/classes/Inventory.ts:200]: I think this is a completely different thing.

File: src/classes/InventoryManager.ts [hunk 4]
```diff
@@ -39,34 +38,14 @@ export default class InventoryManager {
...
+        const matchGeneric = !match && generics ? this.pricelist.getPrice(sku, true, true) : null;
+        match = matchGeneric || match;
         // Amount in inventory should only use generic amount if there is a generic sku
-        const amount =
-            genericCheck && match && genericIndex !== -1
-                ? this.inventory.getAmountOfGenerics(SKU.fromObject(gSku), true)
-                : this.inventory.getAmount(sku, true);
+        const amount = matchGeneric
+            ? this.inventory.getAmountOfGenerics(sku, true)
+            : this.inventory.getAmount(sku, true);
```
Comment (idinium96) [inline src/classes/InventoryManager.ts:48]: Why did you remove almost everything here?

File: src/classes/Pricelist.ts [hunk 5]
```diff
@@ -198,112 +204,43 @@ export default class Pricelist extends EventEmitter {
...
         }
 
-        if (onlyEnabled && !match.enabled) {
-            // Item is not enabled
-            return null;
+        if (generics) {
+            const gSku = generics ? sku.replace(/;u\d+/, '') : null;
+            if (this.hasPrice(gSku, onlyEnabled)) {
+                return this.prices[gSku];
+            }
```
Comment (idinium96) [inline src/classes/Pricelist.ts:223]: I think this will not work.

File: src/classes/InventoryManager.ts (hunk 4 above)
Comment (arik123) [reply src/classes/InventoryManager.ts:48]: Because we are no longer dealing with any indexes, so it is really simple

File: src/classes/Pricelist.ts (hunk 5 above)
Comment (arik123) [reply src/classes/Pricelist.ts:223]: What makes you think so ?

File: src/classes/Inventory.ts (hunk 3 above)
Comment (arik123) [reply src/classes/Inventory.ts:200]: before, you were getting ammount of all variations of sku(with all effects)
now we filter all skus that have the same defindex and quality `<defindex>;5`
and get amount for them - so instead of getting amount for all existing effects, we get amount just for all effects in our pricelist.

File: src/classes/Pricelist.ts (hunk 5 above)
Comment (idinium96) [reply src/classes/Pricelist.ts:223]: Sorry for the delay. I will try to come up with my guess later. Thanks btw.

**Comment by arik123** [comment]
Any progress here ?

**Comment by idinium96** [comment]
> Any progress here ?

I am taking a break for v4 development, but yeah I will do some fix on `tf2-schema-2` (https://github.com/TF2Autobot/node-tf2-schema-2/issues/8), and then I will merge `v4-dev` with the current `development` branch and then merge `v4-dev` with this branch and solve any conflict that might appear later.

**Comment by arik123** [comment]
I can do the merges, just let me know whem you are done with tf2-schema-2

**Comment by idinium96** [comment]
> I can do the merges, just let me know whem you are done with tf2-schema-2

Alright, I've merged `development` branch with `v4-dev`. You're clear to merge `v4-dev` with this branch.

**Comment by arik123** [comment]
So I merged it, it seems that there were no major issues, but there are a few lint errors that I don't have a clue on how to fix.

**Comment by arik123** [comment]
There still is problem with one test case, but otherwise it seems to work. I tried to trade with the bot, and it seems fine.
image

**Comment by idinium96** [comment]
Any idea as to why this happened sir @joekiller ?

image

image

**Comment by idinium96** [comment]
I think this is ready to merge. But maybe need to test more for all commands that will modify the pricelist first.
